runtime/runtime/src/lib.rs

```diff
@@ -1434,22 +1502,55 @@ impl Runtime {
                 set_postponed_receipt(&mut state_update, &receipt);
             }
         }
-        if delayed_receipts_indices != DelayedReceiptIndices::default() {
-            set(&mut state_update, TrieKey::DelayedReceiptIndices, &delayed_receipts_indices);
-        }
 
-        for (account_id, _, amount) in validators {
+        for (account_id, _, amount) in self.validators {
+            if !chunk_account_ids.contains(account_id) {
+                continue;
+            }
             let mut account: Account = get_account(&state_update, account_id)
                 .expect("Genesis storage error")
                 .expect("account must exist");
             account.set_locked(*amount);
             set_account(&mut state_update, account_id.clone(), &account);
         }
-        state_update.commit(StateChangeCause::InitialState);
-        let trie_changes = state_update.finalize_genesis().expect("Genesis state update failed");
 
-        let (store_update, state_root) = tries.apply_genesis(trie_changes, shard_id);
-        (store_update, state_root)
+        self.commit(state_update);
+    }
+
+    pub fn apply_delayed_receipts(&mut self) {
+        let mut state_update = self.tries.new_trie_update(self.shard_id, self.current_state_root);
+
+        if self.delayed_receipts_indices != DelayedReceiptIndices::default() {
+            set(&mut state_update, TrieKey::DelayedReceiptIndices, &self.delayed_receipts_indices);
+            self.commit(state_update);
+        }
+    }
+
+    pub fn apply(&mut self, genesis: &Genesis, shard_account_ids: HashSet<AccountId>) -> StateRoot {
+        for chunk_account_ids in
+            shard_account_ids.into_iter().collect::<Vec<AccountId>>().chunks(120000)
+        {
+            self.apply_chunk(genesis, HashSet::from_iter(chunk_account_ids));
+        }
+        self.apply_delayed_receipts();
+        self.current_state_root
+    }
+}
+
+impl Runtime {
+    /// Balances are account, publickey, initial_balance, initial_tx_stake
+    pub fn apply_genesis_state(
+        &self,
+        tries: ShardTries,
+        shard_id: ShardId,
+        validators: &[(AccountId, PublicKey, Balance)],
+        genesis: &Genesis,
+        config: &RuntimeConfig,
+        shard_account_ids: HashSet<AccountId>,
+    ) -> StateRoot {
+        let mut genesis_state_applier =
+            GenesisStateApplier::new(tries, shard_id, validators, config);
+        genesis_state_applier.apply(genesis, shard_account_ids)
```

Comment:
Ok, at least it seems that *this* dosen't need to be a stateful object, as we apply all the stuff in one go anyway.

Reply:
I'm not sure if I've done what you wanted, it still looks weird. Also this part of code moved to runtime/runtime/src/genesis.rs